OCPBUGS 4543: Prioritizing deployments over deployment configs (#63036)
openshift/openshift-docs: modules/deployments-comparing-deploymentconfigs.adoc

However, deployment rollouts are driven from a controller manager. The controller manager runs in high availability mode on masters and uses leader election algorithms to value availability over consistency. During a failure it is possible for other masters to act on the same deployment at the same time, but this issue will be reconciled shortly after the failure occurs.

[id="delpoyments-specific-features_{context}"]
== Deployment-specific features

[discrete]
==== Rollover

The deployment process for `Deployment` objects is driven by a controller loop, in contrast to `DeploymentConfig` objects that use deployer pods for every new rollout. This means that the `Deployment` object can have as many active replica sets as possible, and eventually the deployment controller will scale down all old replica sets and scale up the newest one.

> `DeploymentConfig` objects can have at most one deployer pod running, otherwise multiple deployers might conflict when trying to scale up the newest replication controller. Because of this, only two replication controllers can be active at any point in time. Ultimately, this results in faster rapid rollouts for `Deployment` objects.

Author: Thanks, reverted

> `DeploymentConfig` objects can have at most one deployer pod running, otherwise multiple deployers might conflict when trying to scale up the newest replication controller. Because of this, only two replication controllers can be active at any point in time. Ultimately, this results in faster rapid rollouts for `Deployment` objects.

Reviewer: thx!

[discrete]
==== Proportional scaling

Because the deployment controller is the sole source of truth for the sizes of new and old replica sets owned by a `Deployment` object, it can scale ongoing rollouts. Additional replicas are distributed proportionally based on the size of each replica set.

> `DeploymentConfig` objects cannot be scaled when a rollout is ongoing because the controller migtht have issues with the deployer process about the size of the new replication controller.

Reviewer: ```suggestion
`DeploymentConfig` objects cannot be scaled when a rollout is ongoing because the controller  will end up having issues with the deployer process about the size of the new replication controller.
```

It is certain here that problems will arise.

However, deployment rollouts are driven from a controller manager. The controller manager runs in high availability mode on masters and uses leader election algorithms to value availability over consistency. During a failure it is possible for other masters to act on the same deployment at the same time, but this issue will be reconciled shortly after the failure occurs.

[id="delpoyments-specific-features_{context}"]
== Deployment-specific features

[discrete]
==== Rollover

> The deployment process for `Deployment` objects is driven by a controller loop, in contrast to `DeploymentConfig` objects which use deployer pods for every new rollout. This means that the `Deployment` object can have as many active replica sets as possible, and eventually the deployment controller will scale down all old replica sets and scale up the newest one.

Author: +1, I like "that" better too

> `DeploymentConfig` objects can have at most one deployer pod running, otherwise multiple deployers end up conflicting while trying to scale up what they think should be the newest replication controller. Because of this, only two replication controllers can be active at any point in time. Ultimately, this translates to faster rapid rollouts for `Deployment` objects.

Author: Agreed that this sentence needs some work. I cut out some of it. @atiratree Can you let me know if this retains an accurate meaning:

```suggestion
`DeploymentConfig` objects can have at most one deployer pod running, otherwise multiple deployers might conflict when trying to scale up the newest replication controller. Because of this, only two replication controllers can be active at any point in time. Ultimately, this translates to faster rapid rollouts for `Deployment` objects.
```

> `DeploymentConfig` objects can have at most one deployer pod running, otherwise multiple deployers end up conflicting while trying to scale up what they think should be the newest replication controller. Because of this, only two replication controllers can be active at any point in time. Ultimately, this translates to faster rapid rollouts for `Deployment` objects.

Author: +1 good suggestion

[discrete]
==== Proportional scaling

Because the deployment controller is the sole source of truth for the sizes of new and old replica sets owned by a `Deployment` object, it is able to scale ongoing rollouts. Additional replicas are distributed proportionally based on the size of each replica set.

`DeploymentConfig` objects cannot be scaled when a rollout is ongoing because the controller will end up having issues with the deployer process about the size of the new replication controller.

[discrete]
==== Pausing mid-rollout

Deployments can be paused at any point in time, meaning you can also pause ongoing rollouts. On the other hand, you cannot pause deployer pods

> currently, so if you try to pause a deployment in the middle of a rollout, the deployer process will not be affected and will continue until it finishes.

Author: Rewrote that paragraph a bit thanks!

==== Custom strategies

> Deployments do not support user-specified custom deployment strategies yet.

Reviewer: Not very likely they will ever get supported, so I would remove the `yet` part.

Author: Removed, thanks!
